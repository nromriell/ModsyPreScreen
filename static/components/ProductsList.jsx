import React, {useEffect, useLayoutEffect} from 'react'
import {connect} from 'react-redux';
import ProductCard from './ProductCard';
import InfiniteScroll from './InfiniteScroll';
import {action_get_remote_products, action_set_scroll_load_target} from "../redux/actions";

import '../style/theme.css'

/**
 * @author
 * Nathan Romriell - 10/16/2019
 *
 * @info
 * Functional Component - ProductsList
 *
 * @description
 * ProductCard Parent utilizing generic InfiniteScroll
 *
 * @redux-props
 * data                   - Product data
 * setLoadTarget($target) - Set the target Y for the next page load based on the loaded elements
 *
 * @Hooks
 * useLayoutEffect        - Used to update the target Y for page load after component layout
 * useEffect              - Used to add/remove resize event listeners to modify the target Y for page load
 *
 */
const ProductsList = (props) => {
    const updateLastItemPosition = () => {
        const lastItem = document.querySelector("#ProductsList:last-child");
        if(lastItem) {
            const lastItemPosition = lastItem.offsetTop + lastItem.clientHeight;
            props.setLoadTarget(lastItemPosition);
        }
    };
    useLayoutEffect(() => {
       updateLastItemPosition();
    });
    useEffect(() => {
       window.addEventListener('resize', updateLastItemPosition);
       return () => {
         window.removeEventListener('resize', updateLastItemPosition);
       };
    });
    const list = props.data.products.map((product) =>
    <ProductCard key={product.sku} url={product.url} name={product.name} description={product.description} price={product.price} image={product.image}/>);
    return <InfiniteScroll
        id="ProductsList"
        className="ProductsList"
        pageLoadAction={action_get_remote_products}
        targetChangeAction={action_set_scroll_load_target}>
        {list}
    </InfiniteScroll>
};

const mapStateToProps = (state) =>
{
    return {data:state.data};
};

const mapDispatchToProps = (dispatch) =>
{
    return {
        setLoadTarget: (target) => dispatch(action_set_scroll_load_target(target))
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(ProductsList)