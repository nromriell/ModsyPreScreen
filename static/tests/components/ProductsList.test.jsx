import React from 'react'
import renderer from 'react-test-renderer'
import {Provider} from 'react-redux'
import ProductsList from '../../components/ProductsList'
import configureStore from 'redux-mock-store'

const mockStore = configureStore([]);

describe('ProductsList Tests', () => {
    let store;
    let component;

    beforeEach(() => {
        store = mockStore({
            data: {loading: false, products:[]},
            scrollTarget: 500
        });
        component = renderer.create(
            <Provider store={store}>
                <ProductsList/>
            </Provider>
        )
    });

    it('Renders Correctly', () => {
        expect(component.toJSON()).toMatchSnapshot();
    });

});