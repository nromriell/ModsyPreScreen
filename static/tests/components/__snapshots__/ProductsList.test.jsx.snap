// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsList Tests Renders Correctly 1`] = `
<div
  className="ProductsList"
  id="ProductsList"
/>
`;
